### PR TITLE
perf(orchestration): parallel multi-query fan-out and wave context sliding window

### PR DESCRIPTION
**Pull request by 2001anshkaushik**
## Summary
This PR hardens the orchestration layer by addressing two critical bottlenecks in agent workflows: I/O latency and context bloat.

**1. Parallel Multi-Query Fan-Out (`DocumentStoreBase`)**
- **The Problem:** Multi-question RAG queries previously executed sequentially, blocking the event loop on every vector DB round-trip.
- **The Fix:** Replaced the sequential `for` loop with a `ThreadPoolExecutor` fan-out. Semantic/Keyword queries now execute concurrently, with a `threading.Lock` serializing the final dictionary merge. This drops latency from `N * p99` to `~p99`. 
- **Resilience:** Includes explicit zero-query tolerance (`if not queries: pass`) to prevent `max_workers=0` crashes, and guarantees in-flight thread cleanup via context managers.

**2. Server-Side Sliding Window (`RocketRideDriver`)**
- **The Problem:** The `waves` context previously grew infinitely unless the LLM cooperatively emitted a `remove` signal, risking context-window overflows.
- **The Fix:** Implemented a dual-strategy `_prune_wave_context()` guard enforcing a sliding window (`context_window_waves`) and a character budget (`wave_context_budget_chars`) before every prompt compilation, ensuring context remains mathematically bounded.

## Type
performance, enhancement

## Testing
- [x] Tests added or updated (26 new tests in `test_document_store_parallel.py`, covering concurrency stress tests and zero-query edge cases)
- [x] Tested locally
- [x] `./builder test` passes cleanly

## Checklist
- [x] Commit messages follow [conventional commits](https://www.conventionalcommits.org/)
- [x] No secrets or credentials included
- [x] 0-blast-radius for single-query pipelines

## Summary by CodeRabbit

* **New Features**
  * Automatic wave-history context pruning with configurable window size and per-window character budget to keep planning context bounded.
  * Parallelized document-store sub-query execution with a capped concurrency limit to speed multi-query searches and preserve correct aggregation.

* **Tests**
  * New test suites covering parallel query fan-out, aggregation, de-duplication, edge cases, concurrency/thread-safety, and context-pruning behavior; includes a test-time stub for an external dependency.

[Review Change Stack](https://app.coderabbit.ai/change-stack/rocketride-org/rocketride-server/pull/869)

### CODE REVIEW

File: nodes/src/nodes/agent_rocketride/rocketride_agent.py [hunk 1]
```diff
@@ -208,6 +279,12 @@ def _run(
             # as context; the full result stays in memory for later peek access.
             results = execute_wave(tool_calls, agent_base=self, context=context, wave_name=f'wave-{wave_num}')
             waves.append({'wave_num': wave_num, 'calls': tool_calls, 'results': results})
+
+            # Server-side context budget enforcement.  Runs after every wave
+            # append so the next planning prompt never exceeds the configured
+            # character budget, regardless of LLM ``remove`` signal behaviour.
+            self._prune_wave_context(waves)
```
[the remark this answers is not in the document]
Comment (2001anshkaushik) [reply nodes/src/nodes/agent_rocketride/rocketride_agent.py]: Refactored to maintain a strict separation of concerns: waves remains the canonical, immutable trace used for final synthesis and observability, while a pruned copy is generated specifically for the planning prompt. This ensures token efficiency without losing historical context for the final answer.

File: packages/ai/tests/ai/common/test_document_store_parallel.py [hunk 2]
```diff
@@ -0,0 +1,648 @@
...
+        elapsed = time.monotonic() - start
+
+        # Sequential would take n_queries * LATENCY; parallel should be < 2× LATENCY.
+        sequential_lower_bound = n_queries * self.SIMULATED_QUERY_LATENCY_S
+        parallel_upper_bound = 2 * self.SIMULATED_QUERY_LATENCY_S
+
+        assert elapsed < parallel_upper_bound, (
+            f'Expected parallel execution in <{parallel_upper_bound:.2f}s '
+            f'(was {elapsed:.3f}s vs sequential lower bound {sequential_lower_bound:.2f}s)'
+        )
```
[the remark this answers is not in the document]
Comment (2001anshkaushik) [reply packages/ai/tests/ai/common/test_document_store_parallel.py]: Replaced the non-deterministic wall-clock timing assertion with a structural threading. Barrier proof. The test now mathematically verifies concurrent execution by requiring all threads to rendezvous simultaneously, making the suite immune to CI runner jitter.

File: packages/ai/tests/ai/common/test_document_store_parallel.py [hunk 3]
```diff
@@ -0,0 +1,648 @@
...
+        context_window_waves: int = 5,
+        wave_context_budget_chars: int = 12_000,
+    ):
+        """Return a minimal stub with the _prune_wave_context algorithm baked in.
+
+        We deliberately avoid importing ``nodes.src.*`` here because the
+        ``packages/ai`` test runner does not have the ``nodes`` package on its
+        ``PYTHONPATH``.  The stub reproduces the algorithm verbatim so the tests
+        validate the logic without requiring the full node import chain.
+        """
```
[the remark this answers is not in the document]
Comment (2001anshkaushik) [reply packages/ai/tests/ai/common/test_document_store_parallel.py]: Agreed. I have removed the local algorithm stub and implemented comprehensive coverage in nodes/test/ that exercises the real RocketRideDriver implementation using the project's established testing patterns (bypassing __init__ via __new__).

**Comment by 2001anshkaushik** [comment]
@coderabbitai resume

**Comment by Rod-Christensen** [comment]
Wave was intended to operate exactly how you implemented it, running multiple subnodes concurrently. However, a problem came up in testing that needs to be fixed in the c++ engine before we can merge. It is a corner case, but it is pretty serious..

Most python nodes assume the sequential architecture. However, this makes them parallel which has the capability to corrupt "temporary" data that they are holding. The solution is, in the C++ engine, we need to ensure that no node is shared between wave partners, or lock a mutex so a node runs to completion prior to returning. Either way, it requires a backend change. Neither is too terfibly difficuly.

**Review by dsapandora — CHANGES_REQUESTED** [review]
This PR should not merge yet. The wave feature is implemented and scoped correctly at the store layer — fan-out via ThreadPoolExecutor in `_queryDocuments` with a properly-scoped merge lock, and `_prune_wave_context()` returning a fresh copy per planning iteration instead of mutating `waves` in place (fixing CodeRabbit's line-286 flag while keeping `trace['waves']` and the synthesis fallback intact). The Barrier-based concurrency test is a solid replacement for a flaky timing check, and the invariant tests around 'master list never mutated' are the right shape for this kind of change.

However, testing surfaced a serious corner case that blocks this PR: most Python nodes were written assuming the engine executes sequentially, and several hold temporary/mutable state during execution. Running wave partners concurrently — which is exactly what this PR correctly implements — exposes those nodes to being shared across concurrent wave partners, which can corrupt that temporary state. This is a real regression introduced by turning previously-sequential execution parallel, and it isn't something the current Python-side tests exercise: the concurrency proof and invariant tests here cover the document store's merge safety, not node-level state isolation.

The fix has to land in the C++ engine, not in this PR's Python code: either (a) ensure no node instance is shared between wave partners (give each wave its own node instances), or (b) hold a per-node mutex so a node runs to completion before it can be re-entered by another wave. Either approach looks workable and shouldn't be a heavy lift, but it's a backend change and a prerequisite for merging wave-parallel execution safely — without it, this corner case can produce corrupted results in production under the exact concurrency this PR enables. New tests exercising node-level state isolation (or mutex serialization) should land alongside whichever fix is chosen, since the current suite doesn't cover this failure mode on any platform.

Once that's addressed, the remaining items are minor and non-blocking:

- The SEMANTIC and KEYWORD fan-out blocks in `store.py` are near-identical; worth a shared helper if this file is touched again.
- The 'Defect 3' framing in `test_document_store_parallel.py` (GET allegedly unreachable via the KEYWORD `elif`) isn't an actual bug before or after this change — `type` is a single enum value, so the branch was always correctly reachable regardless of query count. Worth tightening that docstring.
- `_prune_wave_context` only sums `results[].summary` lengths, not the `calls` payload, so a wave with large tool-call args could still blow past the intended context budget. Worth watching if context-overflow issues resurface.

Given this has apparently been sitting since around May 26, prioritizing the C++ side fix quickly would help — otherwise this is a candidate to close and reopen once the engine change is ready, rather than left open indefinitely.

**Comment by charliegillet** [comment]
Closing per the close-and-reopen suggestion in review rather than leaving this parked. The design was endorsed by both reviewers and the feature is still wanted, but it is hard-blocked on engine-side wave-partner node isolation (per-wave instances or a per-node mutex), which needs to land first and is not yet tracked anywhere. The store refactor in #1663 has also since replaced store.py, so the branch no longer applies. Plan: track the engine prerequisite separately, split off the wave-context pruning half (which does not depend on the concurrency fix) as a small standalone PR, and remake the multi-query fan-out against the new store/ package once the engine work lands.
